airgun/widgets.py

```diff
@@ -489,39 +513,58 @@ class CustomParameter(Widget):
         //textarea[@placeholder='Value']
 
     """
-    add_new_value = Text(".//a[contains(text(),'+ Add Parameter')]")
-    new_parameter_name = TextInput(
-        locator=".//input[@placeholder='Name' and not(@value)]")
-    new_parameter_value = TextInput(
-        locator=".//textarea[@placeholder='Value' and not(text())]")
-    NAMES = (
-        ".//tr[contains(@id, 'os_parameter')]/td/input[@placeholder='Name']")
-    VALUE = (
-        ".//table[contains(@id, 'parameters')]//tr"
-        "/td[input[contains(@id, 'name')][contains(@value, '{}')]]"
-        "/following-sibling::td//textarea"
-    )
+    add_new_value = Text("..//a[contains(text(),'+ Add Parameter')]")
+
+    def __init__(self, *args, **kwargs):
+        """Supports initialization via ``locator=`` or ``id=``"""
+        locator = kwargs.get('locator')
+        id = kwargs.pop('id')
+        if locator and id or not locator and not id:
+            raise ValueError('Please specify either locator or id')
+        kwargs['locator'] = locator or ".//table[@id='{}']".format(id)
+
+        column_widgets = {
+            'Name': TextInput(locator=".//input[@placeholder='Name']"),
+            'Value': TextInput(locator=".//textarea[@placeholder='Value']"),
+            'Actions': Text(locator=".//a[@title='Remove Parameter']")
+        }
+        super(CustomParameter, self).__init__(
+            *args, column_widgets=column_widgets, **kwargs)
 
     def read(self):
         """Return a list of dictionaries. Each dictionary consists of name and
         value parameters
         """
         parameters = []
-        for item in self.browser.elements(self.NAMES):
-            name = self.browser.get_attribute('value', item)
-            value = self.browser.text(self.VALUE.format(name))
+        for row in self.rows():
```

Comment:
if you actually transforming that widget to table, so see no reason for extra logic. It should be simple `table.read()`

Reply:
I doing it this way so that I don't alter the return format (a list of dictionaries with name/value keys). Do we want to change the returned content?

Reply:
probably, you are right. For that specific widget we only interested in these two fields values